src/aspire/abinitio/commonline_c3_c4.py

```diff
@@ -791,31 +791,51 @@ def _signs_times_v(self, vijs, vec):
     @staticmethod
     def _complete_third_row_to_rot(r3):
         """
-        Construct a rotation matrix whose third row is equal to the given row vector.
+        Construct rotation matrices whose third rows are equal to the given row vectors.
         For vector r3 = [a, b, c], where [a, b, c] != [0, 0, 1], we return the matrix
         with rows r1, r2, r3, given by:
 
         r1 = 1/sqrt(a^2 + b^2)[b, -a, 0],
         r2 = 1/sqrt(a^2 + b^2)[ac, bc, -(a^2 + b^2)].
 
-        :param r3: A 1x3 vector of norm 1.
-        :return: A 3x3 rotation matrix whose third row is r3.
+        :param r3: A nx3 array where each row vector has norm 1.
+        :return: An nx3x3 array of rotation matrices whose third rows are r3.
         """
 
+        # Handle singleton vector.
+        singleton = False
+        if r3.shape == (3,):
+            r3 = np.expand_dims(r3, axis=0)
+            singleton = True
+
+        # Initialize output rotation matrices.
+        rots = np.zeros((len(r3), 3, 3), dtype=r3.dtype)
+
+        # Populate 3rd rows.
+        rots[:, 2] = r3
+
+        # Mask for third rows that coincide with the z-axis.
+        mask = np.linalg.norm(r3 - [0, 0, 1], axis=1) < 1e-5
```

Comment:
Easier on the eyes invert this mask so the most used case is the positive case.

Reply:
Done.